### PR TITLE
feat: nanoGPT implementation using Reactant

### PR DESCRIPTION
**Pull request by avik-pal**
### Needs

- [ ] `NNlib.gather` support in Reactant
- [ ] `NNlib.make_causal_mask`: https://github.com/EnzymeAD/Reactant.jl/pull/249

### CODE REVIEW

File: examples/NanoGPT/main.jl [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+    alphabet, trainX, trainY, testX, testY = get_nanogpt_data(; sequence_length, test_split)
+
+    @printf "[Info] Alphabet size: %d\n" length(alphabet)
+    @printf "[Info] Training size: %d sequences.\n" size(trainX, 2)
+    @printf "[Info] Testing  size: %d sequences.\n\n" size(testX, 2)
+
+    train_loader = DataLoader(
+        (trainX, trainY); batchsize, shuffle=true, parallel=true
+    ) |> dev
```
Comment (Vaibhavdixit02) [inline examples/NanoGPT/main.jl:188]:
```suggestion
    ) .|> dev
```

File: examples/NanoGPT/main.jl [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+            time_taken = time() - start_time
+
+            if iter % 100 == 0
+                @printf "[Train] Epoch %3d\tIteration %6d\tLoss %.8e\tTime per \
+                         Iteration %0.5f\n" epoch iter loss time_taken
+            end
+        end
+
+        test_loss = loss_fn(
+            Array(first(model_compiled(testX, ps, Lux.testmode(st)))), testY
```
Comment (Vaibhavdixit02) [inline examples/NanoGPT/main.jl:232]:
```suggestion
            first(model_compiled(testX, ps, Lux.testmode(st))), testY
```

File: examples/NanoGPT/main.jl (hunk 1 above)
Comment (avik-pal) [reply examples/NanoGPT/main.jl:188]: why is this needed? broadcasting over the dataloader will avoid constructing the DeviceIterator

File: examples/NanoGPT/main.jl (hunk 2 above)
Comment (avik-pal) [reply examples/NanoGPT/main.jl:232]: We need to compile the loss_fn for the non-`Array` version to work

File: examples/NanoGPT/main.jl (hunk 1 above)
Comment (Vaibhavdixit02) [reply examples/NanoGPT/main.jl:188]: Atleast, on GPU without this the labels weren't being transferred to the device

File: examples/NanoGPT/main.jl (hunk 2 above)
Comment (Vaibhavdixit02) [reply examples/NanoGPT/main.jl:232]: Ah yeah this might not be needed for your case here
